railties/lib/rails/generators/actions.rb

```diff
@@ -345,6 +344,13 @@ def with_indentation(&block) # :doc:
         ensure
           @indentation -= 1
         end
+
+        # Append string to the +Gemfile+ with a newline if necessary
+        def append_with_newline(str, options = {})
+          gsub_file "Gemfile", /\n?\z/, options do |match|
+            "#{match}\n#{str}\n" unless match.end_with?("\n")
```

Comment:
This looks quite unlike both of my suggestions, and that's fine if it solves the problem, but I don't see how this version works... if the condition is true the block will be nil, which seems unlikely to add `str` to the file. 

If I'm reading it correctly, I think we're either missing test cases, or the conditional is redundant, or something.

Reply:
As beautiful as your suggention looked, they didn't work the way we wanted to so I kind of had to add some changes to them but yeah my code is bit suspicious so let me look into it again.

Reply:
Alright so this is how it works.  The condition`/\n?\z/` matches 2 times when a file ends with newline.

when a file ends with a newline
```ruby
gemfile_text = "source 'https://rubygems.org'\ngit_source(:github) { |repo| \"https://github.com/\#{repo}.git\" }\n\nruby '2.6.1'\n\ngem 'rails', path: \"/Users/masakazutakewaka/workspace/rails\"\n\n# Use sqlite3 as the database for Active Record\ngem 'sqlite3', '~> 1.3', '>= 1.3.6'\n# Use Puma as the app server\ngem 'puma', '~> 3.11'\n# Use SCSS for stylesheets\ngem 'sass-rails', '~> 5.0'\n# Use development version of Webpacker\ngem 'webpacker', github: \"rails/webpacker\"\n\n# Turbolinks makes navigating your web application faster. Read more: https://github.com/turbolinks/turbolinks\ngem 'turbolinks', '~> 5'\n# Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder\ngem 'jbuilder', '~> 2.5'\n# Use Redis adapter to run Action Cable in production\n# gem 'redis', '~> 4.0'\n# Use Active Model has_secure_password\n# gem 'bcrypt', '~> 3.1.7'\n\n# Use Active Storage variant\n# gem 'image_processing', '~> 1.2'\n\ngroup :development, :test do\n  # Call 'byebug' anywhere in the code to stop execution and get a debugger console\n  gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]\nend\n\ngroup :development do\n  # Access an interactive console on exception pages or by calling 'console' anywhere in the code.\n  gem 'web-console', github: 'rails/web-console'\n  gem 'listen', '>= 3.0.5', '< 3.2'\nend\n\ngroup :test do\n  # Adds support for Capybara system testing and selenium driver\n  gem 'capybara', '>= 2.15'\n  gem 'selenium-webdriver'\n  # Easy installation and use of chromedriver to run system tests with Chrome\n  gem 'chromedriver-helper'\nend\n\n# Windows does not include zoneinfo files, so bundle the tzinfo-data gem\ngem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]\n"

class Appendor
  def append_with_newline(text)
    text.gsub(/\n?\z/) do |match|
      p "match : #{match}"
      "#{match}\ngem 'hoge'\n" unless match.end_with?("\n")
    end
  end
end

p Appendor.new.append_with_newline(gemfile_text)


# output
# "match : \n"
# "match : "
# "source 'https://rubygems.org'\ngit_source(:github) { |repo| \"https://github.com/\#{repo}.git\" }\n\nruby '2.6.1'\n\ngem 'rails', path: \"/Users/masakazutakewaka/workspace/rails\"\n\n# Use sqlite3 as the database for Active Record\ngem 'sqlite3', '~> 1.3', '>= 1.3.6'\n# Use Puma as the app server\ngem 'puma', '~> 3.11'\n# Use SCSS for stylesheets\ngem 'sass-rails', '~> 5.0'\n# Use development version of Webpacker\ngem 'webpacker', github: \"rails/webpacker\"\n\n# Turbolinks makes navigating your web application faster. Read more: https://github.com/turbolinks/turbolinks\ngem 'turbolinks', '~> 5'\n# Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder\ngem 'jbuilder', '~> 2.5'\n# Use Redis adapter to run Action Cable in production\n# gem 'redis', '~> 4.0'\n# Use Active Model has_secure_password\n# gem 'bcrypt', '~> 3.1.7'\n\n# Use Active Storage variant\n# gem 'image_processing', '~> 1.2'\n\ngroup :development, :test do\n  # Call 'byebug' anywhere in the code to stop execution and get a debugger console\n  gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]\nend\n\ngroup :development do\n  # Access an interactive console on exception pages or by calling 'console' anywhere in the code.\n  gem 'web-console', github: 'rails/web-console'\n  gem 'listen', '>= 3.0.5', '< 3.2'\nend\n\ngroup :test do\n  # Adds support for Capybara system testing and selenium driver\n  gem 'capybara', '>= 2.15'\n  gem 'selenium-webdriver'\n  # Easy installation and use of chromedriver to run system tests with Chrome\n  gem 'chromedriver-helper'\nend\n\n# Windows does not include zoneinfo files, so bundle the tzinfo-data gem\ngem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]\ngem 'hoge'\n"
```

when a file ends without a newline
```ruby
gemfile_text = "source 'https://rubygems.org'\ngit_source(:github) { |repo| \"https://github.com/\#{repo}.git\" }\n\nruby '2.6.1'\n\ngem 'rails', path: \"/Users/masakazutakewaka/workspace/rails\"\n\n# Use sqlite3 as the database for Active Record\ngem 'sqlite3', '~> 1.3', '>= 1.3.6'\n# Use Puma as the app server\ngem 'puma', '~> 3.11'\n# Use SCSS for stylesheets\ngem 'sass-rails', '~> 5.0'\n# Use development version of Webpacker\ngem 'webpacker', github: \"rails/webpacker\"\n\n# Turbolinks makes navigating your web application faster. Read more: https://github.com/turbolinks/turbolinks\ngem 'turbolinks', '~> 5'\n# Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder\ngem 'jbuilder', '~> 2.5'\n# Use Redis adapter to run Action Cable in production\n# gem 'redis', '~> 4.0'\n# Use Active Model has_secure_password\n# gem 'bcrypt', '~> 3.1.7'\n\n# Use Active Storage variant\n# gem 'image_processing', '~> 1.2'\n\ngroup :development, :test do\n  # Call 'byebug' anywhere in the code to stop execution and get a debugger console\n  gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]\nend\n\ngroup :development do\n  # Access an interactive console on exception pages or by calling 'console' anywhere in the code.\n  gem 'web-console', github: 'rails/web-console'\n  gem 'listen', '>= 3.0.5', '< 3.2'\nend\n\ngroup :test do\n  # Adds support for Capybara system testing and selenium driver\n  gem 'capybara', '>= 2.15'\n  gem 'selenium-webdriver'\n  # Easy installation and use of chromedriver to run system tests with Chrome\n  gem 'chromedriver-helper'\nend\n\n# Windows does not include zoneinfo files, so bundle the tzinfo-data gem\ngem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]"

class Appendor
  def append_with_newline(text)
    text.gsub(/\n?\z/) do |match|
      p "match : #{match}"
      "#{match}\ngem 'hoge'\n" unless match.end_with?("\n")
    end
  end
end

p Appendor.new.append_with_newline(gemfile_text)


# output
# "match : "
# "source 'https://rubygems.org'\ngit_source(:github) { |repo| \"https://github.com/\#{repo}.git\" }\n\nruby '2.6.1'\n\ngem 'rails', path: \"/Users/masakazutakewaka/workspace/rails\"\n\n# Use sqlite3 as the database for Active Record\ngem 'sqlite3', '~> 1.3', '>= 1.3.6'\n# Use Puma as the app server\ngem 'puma', '~> 3.11'\n# Use SCSS for stylesheets\ngem 'sass-rails', '~> 5.0'\n# Use development version of Webpacker\ngem 'webpacker', github: \"rails/webpacker\"\n\n# Turbolinks makes navigating your web application faster. Read more: https://github.com/turbolinks/turbolinks\ngem 'turbolinks', '~> 5'\n# Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder\ngem 'jbuilder', '~> 2.5'\n# Use Redis adapter to run Action Cable in production\n# gem 'redis', '~> 4.0'\n# Use Active Model has_secure_password\n# gem 'bcrypt', '~> 3.1.7'\n\n# Use Active Storage variant\n# gem 'image_processing', '~> 1.2'\n\ngroup :development, :test do\n  # Call 'byebug' anywhere in the code to stop execution and get a debugger console\n  gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]\nend\n\ngroup :development do\n  # Access an interactive console on exception pages or by calling 'console' anywhere in the code.\n  gem 'web-console', github: 'rails/web-console'\n  gem 'listen', '>= 3.0.5', '< 3.2'\nend\n\ngroup :test do\n  # Adds support for Capybara system testing and selenium driver\n  gem 'capybara', '>= 2.15'\n  gem 'selenium-webdriver'\n  # Easy installation and use of chromedriver to run system tests with Chrome\n  gem 'chromedriver-helper'\nend\n\n# Windows does not include zoneinfo files, so bundle the tzinfo-data gem\ngem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]\ngem 'hoge'\n"
```

Reply:
fixed `"#{match}\n#{str}\n"` to `"\n#{str}\n"`

Reply:
@matthewd https://github.com/rails/rails/pull/35285/commits/dbb850a2de517ac810107452ecd9b1590be62f2f 

Hey how does it look like now? 
I think the readability is improved.